packages/runtime/src/execute/expression.ts

```diff
@@ -100,13 +101,30 @@ export const wrapOperation = (
   fn: Operation,
   logger: Logger,
   name: string,
-  immutableState?: boolean
+  immutableState?: boolean,
+  prevName?: string
 ) => {
   return async (state: State) => {
     logger.debug(`Starting operation ${name}`);
     const start = new Date().getTime();
+    if (checkAndClearNullState(state)) {
+      logger.warn(`Operation ${name} might fail!`);
+      logger.warn(
+        `The previous operation ${prevName} didn't return a state. did you forget?`
+      );
+    }
     const newState = immutableState ? clone(state) : state;
-    const result = await fn(newState);
+
+    if (typeof fn !== 'function') {
+      logger.warn(`Are you sure ${name} is an operation?`);
```

Comment:
Logging on this one is tricky! :thinking: 

The "skipped" function is actually executed, right? It's not run as part of the pipeline. So we're in a weird space.

Let's not print a duration. Let's also print just one debug line and one warning.

This will get better when I've got source mapping plugged in so let's not worry about it too much

So can I suggest:
```
logger.warn(`WARNING: operation ${name} is not a valid operation! Make sure to only use functions provided by an adaptor at the top scope.`)
logger.debug(`Skipping operation {$} as invalid`)
```

This isn't very well explained by the docs so I'll add some more docs and then come back and put a link in the error later.

Reply:
Actually I'm super tempted to throw in this case and refuse to run :thinking: 

If the user has done this, then their code is not executing like they think it is. Functions will not execute properly, or they'll run out of order.

It's relatively OK if it's just a console.log - it'll log nonsense but nothing will blow up. But it's extremely unlikely that the user will get the behaviour they want, so maybe we need to force them to remove the offending line.

Let me think about this. Maybe I'll accept the logging for this PR but I'll raise an issue to think about being stricter.

Reply:
Yup that's right. 
1. just as you said, we are actually supposed to error and abort. Because such failure probably will affect the subsequent operations.
2. I was then thinking, the only things that will fail here are top-level blocks of code that actually aren't imported functions from our adaptors.

Reply:
I'm removing the function check out of this PR. probably needs a new one.

Reply:
will be in a different PR!